### PR TITLE
Change default initializer of Mesh()

### PR DESCRIPTION
**Pull request by aruzzo**
Hi, I wanted to introduce myself quickly as this is my first PR to PyNE. I'm Anthony Ruzzo, a current undergrad student in Nuclear Engineering at the University of Illinois at Urbana-Champaign. I'm working under @katyhuff to help address different issues with PyNE. 

This PR addresses issue #1095:

- Default initialization of a mesh, `m=pyne.mesh.Mesh()` now raises an exception with information on how to create a PyNE mesh properly

### CODE REVIEW

File: pyne/mesh.py [hunk 1]
```diff
@@ -704,6 +704,18 @@ def __init__(self, mesh=None, structured=False,
...
+	structured_set is None and structured_ordering=='xyz' and \
+	mats==():
+	    raise MeshError("Trivial mesh instantiation "
+				"For structured mesh instantiation, need to "
+                                "supply exactly one of the following:\n"
+                                "A. PyMOAB instance\n"
+                                "B. Mesh file\n"
+                                "C. Mesh coordinates\n"
+                                "D. Structured entity set AND PyMOAB instance")
+	    return
```
Comment (bam241) [inline pyne/mesh.py]: You are mixing `tabs` and `space` for the indentation. 
We use PEP8 recommendations which recommend to prefer `tabs` over `space` [[see](https://www.python.org/dev/peps/pep-0008/#tabs-or-spaces)].

Moreover `Python3` does not allow tab/space mixing (this is why one of the python3 test is failing).

**Comment by bam241** [comment]
Thank you for putting this In !!!

File: pyne/mesh.py [hunk 2]
```diff
@@ -704,6 +704,18 @@ def __init__(self, mesh=None, structured=False,
             stored in self.dims.
 
         """
+	# if Mesh is made and no parameters passed, raise MeshError
+	if mesh is None and not structured and structured_coords is None and \
+	structured_set is None and structured_ordering=='xyz' and \
```
Comment (bam241) [inline pyne/mesh.py]:
```suggestion
	    structured_set is None and structured_ordering=='xyz' and \
```

File: pyne/mesh.py [hunk 3]
```diff
@@ -704,6 +704,18 @@ def __init__(self, mesh=None, structured=False,
...
+	mats==():
```
Comment (bam241) [inline pyne/mesh.py]:
```suggestion
	    mats==():
```

Comment (bam241) [reply pyne/mesh.py]: maybe consider moving the `mats==():` line to the previous one if it fits (<80 char/line)

**Comment by pshriwise** [comment]
Welcome @aruzzo! Thanks for your PR and for addressing this issue.

File: pyne/mesh.py [hunk 4]
```diff
@@ -704,6 +704,17 @@ def __init__(self, mesh=None, structured=False,
...
+        if mesh is None and not structured and structured_coords is None and \
+            structured_set is None and structured_ordering=='xyz' and mats==():
+            raise MeshError("Trivial mesh instantiation "
+                            "For structured mesh instantiation, need to "
+                                "supply exactly one of the following:\n"
+                                "A. PyMOAB instance\n"
+                                "B. Mesh file\n"
+                                "C. Mesh coordinates\n"
+                                "D. Structured entity set AND PyMOAB instance")
+            return
```
Comment (zxkjack123) [inline pyne/mesh.py]:
```suggestion
```

**Review by zxkjack123 — COMMENTED** [review]
Suggest to add input checking for unstructured mesh too.

File: pyne/mesh.py (hunk 4 above)
Comment (gonuke) [reply pyne/mesh.py]: Note that you don't need a return here because you have raised the exception.  The code can never reach this line.

File: pyne/mesh.py [hunk 5]
```diff
@@ -704,6 +704,16 @@ def __init__(self, mesh=None, structured=False,
             stored in self.dims.
 
         """
+        # if Mesh is made and no parameters passed, raise MeshError
+        if mesh is None and not structured and structured_coords is None and \
```
Comment (pshriwise) [inline pyne/mesh.py]: This is a pretty beefy conditional statement and could be kinda hard to follow for future developers (though the comment helps a lot. Can I suggest making a tuple of our default args that will trigger this exception? Something like?

```python
defaults = (None, False, None, None, 'xyz', ())

if (mesh, structured, structured_coords, structured_set, structured_ordering, mats) == defaults:
    raise MeshError(...)
```

Comment (pshriwise) [reply pyne/mesh.py]: (Note that line may need to be wrapped to comply with PEP8)

Comment (katyhuff) [reply pyne/mesh.py]: This seems wise. Are you willing to implement this suggestion @Mongwell ?

File: pyne/mesh.py [hunk 6]
```diff
@@ -704,6 +704,16 @@ def __init__(self, mesh=None, structured=False,
...
+            structured_set is None and structured_ordering=='xyz' and mats==():
+            raise MeshError("Trivial mesh instantiation "
+                            "For structured mesh instantiation, need to "
```
Comment (katyhuff) [inline pyne/mesh.py]:
```suggestion
                            "For structured mesh instantiation, "
```

File: pyne/mesh.py [hunk 7]
```diff
@@ -704,6 +704,16 @@ def __init__(self, mesh=None, structured=False,
             stored in self.dims.
 
         """
+        # if Mesh is made and no parameters passed, raise MeshError
+        if mesh is None and not structured and structured_coords is None and \
+            structured_set is None and structured_ordering=='xyz' and mats==():
+            raise MeshError("Trivial mesh instantiation "
```
Comment (katyhuff) [inline pyne/mesh.py]: Incomplete thought...
Perhaps:
```suggestion
            raise MeshError("Trivial mesh instantiation detected. "
```

File: pyne/mesh.py (hunk 5 above)
Comment (gonuke) [reply pyne/mesh.py]: An alternative may be to add parentheses to the original conditional to make it easier for humans to parse.  I'm not 100% sold on @pshriwise solution because it kind of separates the values from their semantics - it's hard to follow what all the values in `defaults` refer to.

Comment (pshriwise) [reply pyne/mesh.py]: Sure, I can see how that might be the case. It also isn't super flexible if the constructor changes in the future. I personally like the suggestion in [this PR](https://github.com/aruzzo/pyne/pull/1) though it goes perhaps even farther from the more verbose-yet-clear option here. I'm fine with adjusting the formatting here for added clarity and moving on.

File: pyne/mesh.py [hunk 8]
```diff
@@ -705,8 +705,8 @@ def __init__(self, mesh=None, structured=False,
 
         """
         # if Mesh is made and no parameters passed, raise MeshError
-        if mesh is None and not structured and structured_coords is None and \
-            structured_set is None and structured_ordering=='xyz' and mats==():
+        if (mesh is None and not structured and structured_coords is None and \
+            structured_set is None and structured_ordering=='xyz' and mats==()):
```
Comment (gonuke) [inline pyne/mesh.py]: My recommendation for parentheses was to group each of the “clauses”:

```
if (mesh is None) and (not structured) and (structured_coords is None) and \
     (structured_sets is None) and (structured_ordering=‘xyz’) and (mats==()):
```

File: pyne/mesh.py [hunk 9]
```diff
@@ -704,6 +704,16 @@ def __init__(self, mesh=None, structured=False,
             stored in self.dims.
 
         """
+        # if Mesh is made and no parameters passed, raise MeshError
+        if (mesh is None and not structured and structured_coords is None and \
+            structured_set is None and structured_ordering=='xyz' and mats==()):
```
Comment (gonuke) [inline pyne/mesh.py]: My recommendation for parentheses was to group each of the “clauses”:

```
if (mesh is None) and (not structured) and (structured_coords is None) and \
     (structured_sets is None) and (structured_ordering=‘xyz’) and (mats==()):
```

**Comment by gonuke** [comment]
Also... please add an entry in the `news` - see other recent PRs for examples

File: pyne/mesh.py [hunk 10]
```diff
@@ -704,6 +704,17 @@ def __init__(self, mesh=None, structured=False,
             stored in self.dims.
 
         """
+        # if Mesh is made and no parameters passed, raise MeshError
+        if (mesh is None) and (not structured) and (structured_coords is None) \
+            and (structured_set is None) and (structured_ordering=='xyz') \
```
Comment (pshriwise) [inline pyne/mesh.py]: [PEP8](https://www.python.org/dev/peps/pep-0008/#whitespace-in-expressions-and-statements): We'll want some spaces around the `==` here.

**Review by pshriwise — CHANGES_REQUESTED** [review]
Super small PEP8 comment but other than that I think it's good to go.

**Comment by aruzzo** [comment]
I accidentally force-pushed while making a different PR but I added the spaces around the `==` statements. Everything else should be the same.

**Review by pshriwise — APPROVED** [review]
LGTM @aruzzo. Thanks for staying w/ this!
